File: leetcode/top-interview-questions/easy/array/two-sum.py
class Solution:
  def twoSum(self, nums: List[int], target: int) -> List[int]:
    storage = {}
    for i, num in enumerate(nums):
      storage[target - num] = i
    
    for j, num in enumerate(nums):
      if num not in storage:
        continue
      i = storage[num]
      if i == j:
        continue
      if num + nums[i] == target:
        return [i, j]
    return [-1, -1]

File: leetcode/top-interview-questions/easy/array/two-sum.cpp
class Solution {
 public:
  vector<int> twoSum(vector<int>& nums, int target) {
    // Map every combination of $target - nums[i]$ to i
    // map[9 - 2] = 1
    map<int, int> m;

    for (int i = 0; i < nums.size(); i++) {
      m[target - nums[i]] = i;
    }

    // Check if map[nums[j]] == some i; return [i, j]
    for (int j = 0; j < nums.size(); j++) {
      // cout << j << endl;
      if (m.find(nums[j]) != m.end()) {
        if (m[nums[j]] == j) continue;
        return {m[nums[j]], j};
      }
    }

    return {-1, -1};
  }
};
